### ISSUE TITLE
InChI additional arguments silently ignored

### ISSUE THREAD

**Issue by RaphaelRobidas** [openbabel/openbabel#2526]
When generating an InChI key for a molecule with incomplete valence (e.g., CH3 radical), `openbabel` always generates the InChI key corresponding to the molecule with additional hydrogens, even with the "DoNotAddH" option. The `inchi-1` executable produces the expected output.

## Environment Information
Open Babel version: 3.1.1
Operating system and version: Linux, Debian Bullseye

## Expected Behavior
Using this `.mol` file:
```
Title


  4  3  0  0  0  0  0  0  0  0  0    0
   -3.1837    1.0242    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
   -2.8270    0.0154    0.0000 H   0  0  0  0  0  0  0  0  0  0  0  0
   -2.8270    1.5286   -0.8737 H   0  0  0  0  0  0  0  0  0  0  0  0
   -4.2537    1.0242    0.0000 H   0  0  0  0  0  0  0  0  0  0  0  0
  1  2  1  0  0  0  0
  1  3  1  0  0  0  0
  1  4  1  0  0  0  0
M  END
```
Running `inchi-1 -DoNotAddH methane_radical.mol` yields `InChI=1S/CH3/h1H3` in the output `.txt` file. `openbabel` should also yield this InChI key for the same input and the same option.

## Actual Behavior
However, when using `openbabel`:
```
$ obabel methane_radical.mol -oinchi -xX 'DoNotAddH'                                                                                               
InChI=1S/CH4/h1H4
1 molecule converted

$ obabel methane_radical.mol -oinchi -xX'DoNotAddH'                                                                                                
InChI=1S/CH4/h1H4
1 molecule converted

$ obabel methane_radical.mol -oinchi -xX='DoNotAddH'                                                                                               
InChI=1S/CH4/h1H4
1 molecule converted

$ obabel methane_radical.mol -oinchi -xX='InvalidOption'                                                                                           
InChI=1S/CH4/h1H4
1 molecule converted

$ obabel methane_radical.mol -oinchi -xfdsfs 'InvalidOption'                                                                                       
InChI=1S/CH4/h1H4
1 molecule converted

$ obabel methane_radical.mol -oinchi ---fdsfxfdsfs 'InvalidOption'                                                                                 
InChI=1S/CH4/h1H4
1 molecule converted

$ obabel methane_radical.mol -oinchi -xx 'DoNotAddH'                                                                                               
InChI=1S/CH4/h1H4
1 molecule converted

$ obabel methane_radical.mol -oinchi -xx'DoNotAddH'                                                                                                
InChI=1S/CH4/h1H4
1 molecule converted

$ obabel methane_radical.mol -oinchi --xx'DoNotAddH'                                                                                               
InChI=1S/CH4/h1H4
1 molecule converted

$ obabel methane_radical.mol -oinchi --xx'DoNotAdddfsf'                                                                                            
InChI=1S/CH4/h1H4
1 molecule converted

$ obabel methane_radical.mol -xX 'DoNotAddH' -oinchi                                                                                               
InChI=1S/CH4/h1H4
1 molecule converted

$ obabel methane_radical.mol -xx 'DoNotAddH' -oinchi                                                                                               
InChI=1S/CH4/h1H4
1 molecule converted

$ obabel methane_radical.mol -xx'DoNotAddH' -oinchi                                                                                                
InChI=1S/CH4/h1H4
1 molecule converted
```

So not only is it not giving the expected output, but there is no error message and no useful output.

**Comment by nbehrnd** [comment]
The `.mol` file submitted to openbabel misses an indication that the molecule is a radical (`RAD`); otherwise, the algorithm seems to work fine.  This applies at least for openbabel 3.1.1 as the repositories of Linux Debian 12/bookworm (testing) currently provides (i.e., Mar 29 2022 update).

For example, let the test page of [ChemDraw JS](https://chemdrawdirect.perkinelmer.cloud/js/sample/index.html#) create a methyl radical (there is a button name to structure) with an input of «methyl radical»

![2022-08-13_chemdraw_js](https://user-images.githubusercontent.com/19709084/184505193-d7e5ba2c-83d6-425a-85ab-9e9000eca693.png)

and subsequently, run structure -> get mol to obtain the following file:

```
ACS Document 1996
  ChemDraw08132219512D

  1  0  0  0  0  0  0  0  0  0999 V2000
    0.0000    0.0000    0.0000 C   0  4  0  0  0  0  0  0  0  0  0  0
M  RAD  1   1   2
M  STY  1   1 DAT
M  SLB  1   1   1
M  SAL   1  1   1
M  SDD   1    -1.3654   -1.2329    DA    ALL  1       5
M  SED   1 methyl radical
M  END
```

Pasting this content into `test.mol` and running

```bash
$ obabel test.mol -oinchi
InChI=1S/CH3/h1H3
1 molecule converted
```
does not lead to an additional fourth hydrogen atom.

What appears to be missing in the `.mol` you present to the assignment of an InChI is the statement to retain the molecule as a radical (the line including the keyword `RAD`).  E.g. for a `.mol` file derived from the above eventually including three hydrogen atoms

```bash
$ obabel test.mol -h --gen3d -O test_withH.mol
1 molecule converted
$ cat test_withH.mol 
ACS Document 1996
 OpenBabel081322

  4  3  0  0  0  0  0  0  0  0999 V2000
    1.0126   -0.1536   -0.1280 C   0  0  0  0  0  3  0  0  0  0  0  0
    2.0332    0.1727    0.0263 H   0  0  0  0  0  0  0  0  0  0  0  0
    0.5023   -0.7174    0.6425 H   0  0  0  0  0  0  0  0  0  0  0  0
    0.5022    0.0841   -1.0527 H   0  0  0  0  0  0  0  0  0  0  0  0
  1  2  1  0  0  0  0
  1  3  1  0  0  0  0
  1  4  1  0  0  0  0
M  RAD  1   1   2
M  END
$ obabel test_withH.mol -oinchi
InChI=1S/CH3/h1H3
1 molecule converted
```

To get more familiar how `.mol` (and if you want to process batches of models, the container form of `.sdf`) encodes singlet, doublet, or triplet radicals, I recommend Biovia's documentation (revised edition by 2020) which is public ([link to a .pdf file on archive.org](https://web.archive.org/web/20210219065450/https://discover.3ds.com/sites/default/files/2020-08/biovia_ctfileformats_2020.pdf), or the original [.pdf download from Biovia](https://discover.3ds.com/sites/default/files/2020-08/biovia_ctfileformats_2020.pdf).

--- 

There is an ongoing discussion among the developers of [InChI trust](https://www.inchi-trust.org/) and the developers of FOSS about the current license scheme regarding implementations *computing* InChI in FOSS.  Right of the tip of my hat, I don't recall if OpenBabel is in the fraction «be safe even if it means using an old version» (i.e., corresponding to InChI 1.03 by June 2010), or found a special agreement with InChI trust.  Reference libraries and executables for Linux, Mac, Windows including to contemporary version 1.06 (released December 2020) may be downloaded for free from InChI trust [here](https://www.inchi-trust.org/download-latest-inchi-standard-software/).

**Comment by RaphaelRobidas** [comment]
Thanks for the detailed reply. I understand the source of the issue, but it seems very unpractical. I had never heard of the RAD specification and I don't know a 3D builder that uses it. I don't think anyone is going to go through the trouble of manually specifying radicals, especially with many structures. As a workaround, a unique identifier can be obtained with the `openbabel` InChI + the actual number of hydrogens in the input structure. Not ideal, but relatively simple.

It's unfortunate regarding the potential license issues, but it is true that the license is peculiar.

**Comment by nbehrnd** [comment]
May you please clarify what represents for you a «3D builder»?

*Assuming* this refers to a program to set up a computation in a program like Gaussian, Gamess-UK, Games-US, etc. the state of a molecule as radical is described by the spin of the molecule, independent to assigning the molecule as neutral, or charged.

As an example, [Avogadro](https://avogadro.cc/) may either read a normal closed-shell molecule (like methane) from a `.mol` or `.sdf` or one may build one from which e.g., hydrogens are removed.  With optional extensions you may then export this as as starting point for quantum chemistry calculations as illustrated below:*

![example_avogadro](https://user-images.githubusercontent.com/19709084/184554375-3077e7f3-932e-40dc-b8b7-cc81995097c0.gif)

Some assignments of charges and multiplicities for CH3 above only serve as illustration that it can be set up yet it shows that you can discern between e.g., methyl cation, methyl carbanion, and methy radical.

In this perspective, Avogadro trades similar to interfaces like [gabedit](http://gabedit.sourceforge.net/)

![gabedit](https://user-images.githubusercontent.com/19709084/184554505-6e69ed1e-432b-4771-9996-c4bf568f104d.png)

(image credit: example 7 of [screenshots](https://sites.google.com/site/allouchear/Home/gabedit/screenshots))

or [chemcompute](https://chemcompute.org)

![chemcompute](https://user-images.githubusercontent.com/19709084/184554567-311ae995-edc4-4533-aeb5-654503d08c79.png)

*) In case the optional Avogadro Input Generators plugin (in Avogadro's GUI, Extensions -> Download Plugins; this one requires a working installation of Python) does not yet offer a bridge to your quantum chemical program, you may consider proposing an addition either on [GitHub](https://github.com/OpenChemistry/avogadroapp), or the public [discussion board](https://discuss.avogadro.cc/).

**Comment by RaphaelRobidas** [comment]
By "3D builder", I mean software to construct 3D structures like Gaussview, Avogadro, IQmol, Spartan...

Creating inputs for calculations is not the issue. The issue is precisely that these inputs don't contain the explicit multiplicity of the molecule to prevent openbabel from adding hydrogens despite being instructed not to. We can't really expect all 3D builders to ask for or guess the multiplicity of molecules. However, it seems to me like _openbabel should simply not add hydrogens with the option "DoNotAddH"_. Specifying the multiplicity is more like a workaround.

That being said, I'm not familiar with the codebase, so this might be easier said than done. Nevertheless, it is confusing and can be avoided, as demonstrated by the official `inchi` software.

**Comment by baoilleach** [comment]
You misunderstand. OB is not adding (or removing) anything. That MOL file represents methane; you have three explicit hydrogens and one implicit. It's more common to have either all explicit or all implicit hydrogens. Where did this file come from?

The option does nothing because OB does not pass the MOL file to the InChI codebase; it passes methane which it has read from the file.

**Comment by RaphaelRobidas** [comment]
I see, I've always considered that everything was explicit unless specified otherwise. That makes sense, although the behaviour of the program is still strange to me:
```
$ obabel ch3.mol -omol -d                                                                                                                                                                                                                                                                                                                          
Title
 OpenBabel08152215243D

  1  0  0  0  0  0  0  0  0  0999 V2000
   -3.1837    1.0242    0.0000 C   0  0  0  0  0  0  0  0  0  0  0  0
M  END
1 molecule converted

$ obabel ch3.mol -oinchi -d                                                                                                                                                                                                                                                                                                                        
InChI=1S/CH4/h1H4
1 molecule converted
```

In any case, generating the InChI from a XYZ file (which has to be considered as an explicit structure?) gives the expected output:

```
$ obabel ch3.mol -O ch3.xyz                                                                                                                                                                                                                                                                                                                        
1 molecule converted

$ cat ch3.xyz                                                                                                                                                                                                                                                                                                                                      
4
Title
C         -3.18370        1.02420        0.00000
H         -2.82700        0.01540        0.00000
H         -2.82700        1.52860       -0.87370
H         -4.25370        1.02420        0.00000

$ obabel ch3.xyz -oinchi                                                                                                                                                                                                                                                                                                                           
==============================
*** Open Babel Warning  in InChI code
  Title :Accepted unusual valence(s): C(3)
InChI=1S/CH3/h1H3
1 molecule converted
```

The MOL file was the output of Gaussview in my case.

**Comment by ghutchis** [comment]
Yeah, I wouldn't trust the `MOL` from Gaussview. I doubt they're setting `M CHG` or `M RAD` correctly.

**Comment by baoilleach** [comment]
The problem with xyz is that we have to guess which atoms are connected and their bond orders. Okay for methane, but maybe not so for others. If you have Python skillz, you could write some code to 'correct' the structure the way you want. When you read in the molecule, you just have a loop over the atoms to set the no of implicit Hs to 0.

**Comment by nbehrnd** [comment]
Not only that the typical .xyz file doesn't mention bond orders, if e.g., derived from a .mol file, there is no information retained if/how the molecule was charged which could help in a (re)assignment e.g., with [xyz2mol](https://github.com/jensengroup/xyz2mol) by the Jensen group.  Maybe Anders Christensen's [check-sdf](https://github.com/andersx/check-sdf) could help here.

**Comment by RaphaelRobidas** [comment]
I don't think that the lack of explicit connectivity is a huge issue here, it can be guessed quite accurately from the interatomic distances. As long as there aren't any clashes (non-bonded atoms at a distance near the sum of their covalent radii), the bonds are trivial to identify. After all, it's not like Gaussview or Avogadro have some special information about the bonding of the molecule; it outputs the bonds that the user created, which do not have to be chemically reasonable (e.g., bonded atoms very far away or non-bonded atoms very close). At that point, it is debatable whether the correct connectivity if the one chosen by the user or the one that would be computed using DFT/WFT.

Another option is to use a cheap method like GFN2-xTB to calculate the effective bonds and charges in the molecule, giving quite sensible connectivities in `MOL` format. But then, again, it does not seem to set `M CHG` or `M RAD` either, so we're back to the same issue. Are the charges/radicals really necessary to generate the InChI? At least for `CH3`, the standard InChI key is the same for the neutral radical and for the cation.

**Comment by nbehrnd** [comment]
The InChI string is a layered representation of the chemical structure; it contains at least the Hill formula and connectivity of the backbone atoms (in case of organic molecules, C, N, O, P, S, etc. but not H) are read from the .sdf/.mol file.  *If* the structure described is charged, there will be an additional layer e.g., `q+1` (for single positive), or `q-1` (for single negative), too.   Thus,

+ processing a radical like the methyl radical yields

```shell
$ obabel -:"[CH3]" -oinchi
==============================
*** Open Babel Warning  in InChI code
  #1 :Accepted unusual valence(s): C(3)
InChI=1S/CH3/h1H3
1 molecule converted
```

+ while the methyl cation yields
```shell
$ obabel -:"[CH3+]" -oinchi
InChI=1S/CH3/h1H3/q+1
1 molecule converted
```

+ and the methyl carbanion
```shell
$ obabel -:"[CH3-]" -oinchi
InChI=1S/CH3/h1H3/q-1
1 molecule converted
```

These three InChI strings are dissimilar of each other.  If there would be stereogenic centre in the molecule structure (as in mandelic acid, tartaric acid), there would be an additional layer describing this property, too. 

The `1S/` indicates that the string is about assigning *a standard InChI* of version 1.x.  This is a necessity e.g. when to discern tautomers like 2-hydroxy pyridine vs. the pyridone where only a fixed hydrogen-layer yields different InChI (and subsequently different InChIKey).  To illustrate on this very example:

```shell
$ obabel -:"Oc1ccccn1" -oinchi
InChI=1S/C5H5NO/c7-5-3-1-2-4-6-5/h1-4H,(H,6,7)
1 molecule converted
$ obabel -:"O=c1cccc[nH]1" -oinchi
InChI=1S/C5H5NO/c7-5-3-1-2-4-6-5/h1-4H,(H,6,7)
1 molecule converted
$ obabel -:"Oc1ccccn1" -oinchi -xXFixedH
InChI=1/C5H5NO/c7-5-3-1-2-4-6-5/h1-4H,(H,6,7)/f/h7H
1 molecule converted
$ obabel -:"O=c1cccc[nH]1" -oinchi -xXFixedH
InChI=1/C5H5NO/c7-5-3-1-2-4-6-5/h1-4H,(H,6,7)/f/h6H
1 molecule converted
```

[InChI trust](https://www.inchi-trust.org/) compiled some information about and around the notation.  For morphine, the concept is illustrated by the [first appetizer video](https://youtu.be/rAnJ5toz26c) around 01:30 mm:ss.

**Comment by RaphaelRobidas** [comment]
In absence of charge, why not create the InChI with an undefined charge layer? It's good enough far most uses and that's what the official `inchi` package does. Users that need the charge layer can add the charge specifications.

**Comment by nbehrnd** [comment]
@RaphaelRobidas Contrasting to the mandatory Hill formula, the charge layer only is added to the InChI if the underlying structure is charged.  The automatic charge recognition is embedded in the algorithm to yield the standard InChI.  To best of my knowledge, this is not a parameter InChI trust's reference executables currently (version 1.6 by December 2021) offer to toggle off/relax as they do for the perception of stereochemistry.  (And if engaged, this no longer yields the standard InChI `1S/...`, but one non-standard InChI `1/...`.  Well, one .or. multiple options may be engaged simultaneously to yield *a* non-standard InChI; so *standard* is *agreed convention*.)

On the other hand, there is ongoing development of InChI /per se/ as well as of applications engaging InChI.  Past  conferences ([example-1](https://cactus.nci.nih.gov/presentations/NIHInChI_2021-03/NIHInChI.html), [example-2](https://www.inchi-trust.org/decks-open-inchi-days-2022/)) and future events [scheduled](https://www.inchi-trust.org/conferences-meetings-of-interest/) illustrate this as well as InChI trust's section [open education resources](https://www.inchi-trust.org/oer/).  If you identify a spot where the InChI algorithm may be improved further, a post on their public [mailing list](https://sourceforge.net/p/inchi/mailman/inchi-discuss/) (or joining a working group) may be more effective then here on OpenBabel -- because InChI should not depend on the program used to assign them (OpenBabel, RDKit, ChemDraw, ChemDoodle, ACD ChemSketch, DataWarrior, ChemAxon Marvin, etc.).
